packages/gatsby-source-graphql/src/batching/merge-queries.ts

```diff
@@ -0,0 +1,410 @@
+  FragmentSpreadNode,
+  FragmentDefinitionNode,
+  InlineFragmentNode,
+  FieldNode,
+  NameNode,
+  OperationDefinitionNode,
+  Visitor,
+  ASTKindToNode,
+  VariableDefinitionNode,
+  DirectiveNode,
+} from "graphql"
+import _ from "lodash"
+
+export interface IQuery {
+  query: DocumentNode
+  variables: object
+}
+
+export interface IQueryResult {
+  data: object
+}
+
+const Prefix = {
+  create: (index: number): string => `gatsby${index}_`,
+  parseKey: (prefixedKey: string): { index: number; originalKey: string } => {
+    const match = /^gatsby([\d]+)_(.*)$/.exec(prefixedKey)
+    if (!match || match.length !== 3 || isNaN(Number(match[1])) || !match[2]) {
+      throw new Error(`Unexpected data key: ${prefixedKey}`)
+    }
+    return { index: Number(match[1]), originalKey: match[2] }
+  },
+}
+
+/**
+ * Merge multiple queries into a single query in such a way that query results
+ * can be split and transformed as if they were obtained by running original queries.
+ *
+ * Merging algorithm involves several transformations:
+ *  1. Replace top-level fragment spreads with inline fragments (... on Query {})
+ *  2. Add unique aliases to all top-level query fields (including those on inline fragments)
+ *  3. Prefix all variable definitions and variable usages
+ *  4. Prefix names (and spreads) of fragments with variables
+ *  5. Dedupe repeating fragments
+ *
+ * i.e transform:
+ *   [
+ *     `query Foo($id: ID!) { foo, bar(id: $id), ...FooQuery }
+ *     fragment FooQuery on Query { baz }`,
+ *
+ *    `query Bar($id: ID!) { foo: baz, bar(id: $id), ... on Query { baz } }`
+ *   ]
+ * to:
+ *   query (
+ *     $gatsby1_id: ID!
+ *     $gatsby2_id: ID!
+ *   ) {
+ *     gatsby1_foo: foo,
+ *     gatsby1_bar: bar(id: $gatsby1_id)
+ *     ... on Query @originalFragment(name: "FooQuery") {
+ *       gatsby1_baz: baz
+ *     }
+ *     gatsby2_foo: baz
+ *     gatsby2_bar: bar(id: $gatsby2_id)
+ *     ... on Query {
+ *       gatsby2_baz: baz
+ *     }
+ *   }
+ *   fragment FooQuery on Query { baz }
+ */
+export function merge(queries: ReadonlyArray<IQuery>): IQuery {
+  const mergedVariables: object = {}
+  const mergedVariableDefinitions: VariableDefinitionNode[] = []
+  const mergedSelections: SelectionNode[] = []
+  const mergedFragmentMap: Map<string, FragmentDefinitionNode> = new Map()
+
+  queries.forEach((query: IQuery, index: number) => {
+    const prefixedQuery = prefixQueryParts(Prefix.create(index), query)
+
+    prefixedQuery.query.definitions.forEach(def => {
+      if (isQueryDefinition(def)) {
+        mergedSelections.push(...def.selectionSet.selections)
+        mergedVariableDefinitions.push(...(def.variableDefinitions ?? []))
+      }
+      if (isFragmentDefinition(def)) {
+        // Theoretically it is possible to have fragments with the same name but different content
+        // in different queries. But Gatsby validation doesn't allow this and we rely on this here
+        mergedFragmentMap.set(def.name.value, def)
```

Comment:
This is probably okay, but *might* break if people use it in gatsby-node and have same fragments. I'm okay with adding a comment about that particular potential (unlikely) breakage.

Reply:
Added this example to the comment